theseus/optimizer/nonlinear/nonlinear_least_squares.py

```diff
@@ -4,18 +4,83 @@
 # LICENSE file in the root directory of this source tree.
 
 import abc
-from typing import Any, Dict, Optional, Type
+import warnings
+from enum import Enum
+from typing import Any, Callable, Dict, NoReturn, Optional, Tuple, Type, Union
 
+import numpy as np
 import torch
 
 from theseus.core import Objective
-from theseus.optimizer import Linearization
-from theseus.optimizer.linear import LinearSolver
-from theseus.optimizer.linear.dense_solver import CholeskyDenseSolver
+from theseus.optimizer import Linearization, OptimizerInfo
+from theseus.optimizer.linear import (
+    LinearSolver,
+    CholeskyDenseSolver,
+    LUCudaSparseSolver,
+)
+from .nonlinear_optimizer import (
+    NonlinearOptimizer,
+    NonlinearOptimizerInfo,
+    NonlinearOptimizerStatus,
+)
 
-from .nonlinear_optimizer import NonlinearOptimizer
 
+class BackwardMode(Enum):
+    UNROLL = 0
+    IMPLICIT = 1
+    TRUNCATED = 2
+    DLM = 3
```

Comment:
Should BackwardModes go to the base Optimizer class level? Any derived optimizers can then specify which modes they currently support and also have their custom implementation to handle that specific class of optimization problems.